src/subsplit_dag.cpp

```diff
@@ -802,67 +832,51 @@ SizeVector SubsplitDAG::BuildEdgeReindexer(const size_t prev_edge_count) {
-    for (size_t idx = new_idx; idx < edge_idx; idx++) {
-      // Every idx after the new edge's position is shifted by 1.
-      if (edge_reindexer.at(idx) >= new_idx) {
-        edge_reindexer.at(idx) = edge_reindexer.at(idx) + 1;
-      }
-    }
+    const size_t new_idx = edge_reindexer.at(idx_range.second);
+    Reindexer::ReassignAndShift(edge_reindexer, edge_idx, new_idx);
   }
   return edge_reindexer;
 }
 
-void SubsplitDAG::UpdateNodeIds(const SizeVector node_reindexer) {
-  if (node_reindexer != Reindexer::IdentityReindexer(NodeCount())) {
-    // Update `dag_nodes_`.
-    auto dag_nodes_copy = std::move(dag_nodes_); // Copies dag_nodes_ because of unique_ptrs.
-    dag_nodes_ = Reindexer::Reindex(dag_nodes_copy, node_reindexer);
-    // Update each node's id and leafward/rootward ids.
-    for (int id = 0; id < NodeCount(); id++) {
-      GetDAGNode(id)->UpdateIds(node_reindexer);
-    }
-    // Update `subsplit_to_id_`.
-    for (const auto &[subsplit, id] : subsplit_to_id_) {
-      subsplit_to_id_.at(subsplit) = node_reindexer.at(id);
-    }
-    // Update `dag_edges_`.
-    auto dag_edges_copy = std::move(dag_edges_);
-    for (const auto &[node_pair, idx] : dag_edges_copy) {
-      SafeInsert(
-          dag_edges_,
-          {node_reindexer.at(node_pair.first), node_reindexer.at(node_pair.second)},
-          idx);
-    }
+void SubsplitDAG::RemapNodeIds(const SizeVector &node_reindexer) {
+  // Update `dag_nodes_`.
+  auto dag_nodes_copy = Reindexer::Reindex(dag_nodes_, node_reindexer);
+  dag_nodes_.swap(dag_nodes_copy);
+  // Update each node's id and leafward/rootward ids.
+  for (int id = 0; id < NodeCount(); id++) {
+    GetDAGNode(id)->RemapNodeIds(node_reindexer);
+  }
+  // Update `subsplit_to_id_`.
+  for (const auto &[subsplit, id] : subsplit_to_id_) {
+    subsplit_to_id_.at(subsplit) = node_reindexer.at(id);
+  }
+  // Update `dag_edges_`.
+  std::map<SizePair, size_t> dag_edges_copy;
+  for (const auto &[node_pair, idx] : dag_edges_) {
+    SafeInsert(
+        dag_edges_copy,
+        {node_reindexer.at(node_pair.first), node_reindexer.at(node_pair.second)}, idx);
   }
+  dag_edges_.swap(dag_edges_copy);
 }
 
-void SubsplitDAG::UpdateEdgeIdxs(const SizeVector edge_reindexer) {
-  if (edge_reindexer != Reindexer::IdentityReindexer(GPCSPCountWithFakeSubsplits())) {
-    // Update `dag_edges_`.
-    for (const auto &[node_pair, edge_idx] : dag_edges_) {
-      dag_edges_[node_pair] = edge_reindexer[edge_idx];
-    }
-    // Update `parent_to_range_`.
-    for (const auto &[subsplit, idx_range] : parent_to_range_) {
-      parent_to_range_[subsplit] = {edge_reindexer[idx_range.first],
-                                    edge_reindexer[idx_range.second]};
-    }
+void SubsplitDAG::RemapEdgeIdxs(const SizeVector &edge_reindexer) {
+  // Update `dag_edges_`.
+  for (const auto &[node_pair, edge_idx] : dag_edges_) {
+    dag_edges_.at(node_pair) = edge_reindexer.at(edge_idx);
```

Comment:
I wasn't really sure how/if this worked. We're modifying the thing we are looping over! It looks like it looks like it requires a little thought.

See https://stackoverflow.com/questions/8234779/how-to-remove-from-a-map-while-iterating-it ... perhaps the other links there, and either provide evidence that what we're doing is fine or make whatever modifications are necessary.

Reply:
Note this happens just below as well.

Reply:
Dave and I talked, and we both think that this will work because we are only modifying the values in each key-value pair and are not deleting/adding any keys.

Reply:
Great! Can you provide a link describing what makes it OK?